### PR TITLE
SIMD-0153: ZK ElGamal Proof program

### PR DESCRIPTION
**Pull request by samkim-crypto**
## Summary

Deprecate the existing ZK Token Proof program that is tailored for the SPL Token program and replace it with a more general zero-knowledge proof program called the ZK ElGamal Proof program.

## Motivation

The existing native ZK Token Proof program contains logic that is quite specialized for the SPL Token program. This program is a native built-in program that is part of the validator. These type of programs should generally not favor any specific application or program like the SPL Token program.

This document proposes to deprecate the existing ZK Token Proof program and replace it with a new program that is independent of any specific application.

### CODE REVIEW

**Review by 0x0ece — APPROVED** [review]

**Comment by 0x0ece** [comment]
LGTM, FD will support the new program

File: proposals/0153-elgamal-proof-program.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+## Motivation
+
+The existing native ZK Token Proof program contains logic that is
+quite specialized for the SPL Token program. For example, the program contains
+instructions `VerifyTransfer`, `VerifyTransferWithFee`, and `VerifyWithdraw`
+that verify zero-knowledge proofs that are tailor made for `Transfer` and
+`Withdraw` instructions in the SPL Token confidential transfer extension.
+
+The ZK Token Proof program is a native built-in program that is part of the
+validator client. These type of programs should generally not favor any specific
```
Comment (joncinque) [inline proposals/0153-elgamal-proof-program.md:30]: nit typo
```suggestion
validator client. These types of programs should generally not favor any specific
```

File: proposals/0153-elgamal-proof-program.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+any confidential transfers. Nevertheless, including these type of privacy related
+proof verification logic in the validator client can provide unnecessary legal
+burden to the maintainers of the Solana validator clients.
+
+In this document, we propose that we deprecate the existing ZK Token Proof
+program and replace it with a more general ZK ElGamal Proof program that is
+application independent. The new ZK ElGamal Proof program inherits parts of the
+ZK Token Proof program that is independent of any specific application like
+the logic to verify the validity of a public key or range of the encrypted in an
+ElGamal ciphertext. It leaves out parts of the logic that is specific
```
Comment (joncinque) [inline proposals/0153-elgamal-proof-program.md:44]: nit typo
```suggestion
ElGamal ciphertext. It leaves out parts of the logic that are specific
```

File: proposals/0153-elgamal-proof-program.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+    ///
+    /// This instruction will be left unchanged.
+    VerifyBatchedGroupedCiphertext3HandlesValidity,
+}
+```
+
+The implementation of the instructions that are not removed will remain the
+same.
+
+Instead of the modifying the existing ZK Token Proof program into ZK ElGamal
```
Comment (joncinque) [inline proposals/0153-elgamal-proof-program.md:216]: nit typo
```suggestion
Instead of modifying the existing ZK Token Proof program into the ZK ElGamal
```

File: proposals/0153-elgamal-proof-program.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+Proof program, we propose to simply remove the ZK Token Proof program altogether
+and add a new built-in ZK ElGamal Proof program. The existing ZK Token Proof
+program is not yet activated on any of the clusters.
+
+## Impact
+
+The existing ZK Token Proof program will be deprecated and removed. A new ZK
+ElGamal Proof program will be added to the list of built-in programs.
```
Comment (joncinque) [inline proposals/0153-elgamal-proof-program.md:224]: Can you add the address of the new proof program?

File: proposals/0153-elgamal-proof-program.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+## Detailed Design
+
+The new ZK ElGamal Proof program contains a strict subset of the proof
+verification instructions in the original ZK Token Proof program. We list the
+current list of the instructions in the ZK Token Proof program and denote
+whether they are included, removed, or renamed in the new ZK ElGamal Proof
+program.
+
+```rust
+pub enum ProofInstruction {
```
Comment (joncinque) [inline proposals/0153-elgamal-proof-program.md:79]: nit: rather than using a rust enum and code, which can make it hard to parse quickly, can you simply list the instruction names and differences? ie

* `CloseContextState`: unchanged
* `VerifyZeroBalance`: renamed to `VerifyZeroCiphertext`

And if you want, you can provide a link to the current ZK Token Proof program instruction code.

**Review by joncinque — COMMENTED** [review]
Looks good to me! Just a few small things to fixup

File: proposals/0153-elgamal-proof-program.md (hunk 5 above)
Comment (samkim-crypto) [reply proposals/0153-elgamal-proof-program.md]: Okay good idea. I removed the rust enum code and just listed the instructions that are to be renamed or removed. I can even remove the short proof descriptions, but I thought maybe I can keep it just for context. I can remove it if you prefer removing it too.

File: proposals/0153-elgamal-proof-program.md (hunk 4 above)
Comment (samkim-crypto) [reply proposals/0153-elgamal-proof-program.md]: Good idea. Done!

File: proposals/0153-elgamal-proof-program.md (hunk 5 above)
Comment (joncinque) [reply proposals/0153-elgamal-proof-program.md]: What you did looks great, thanks!

**Review by joncinque — APPROVED** [review]
Looks great!

**Comment by joncinque** [comment]
Since there haven't been any comments in the past week, and there's approval from both Anza and Firedancer, I will give one more day for any comments before merging.
